### PR TITLE
Clarify Path.GetPathRoot return value when path is effectively empty

### PR DESCRIPTION
**Pull request by carlossanlop**
As discussed here: https://github.com/dotnet/corefx/issues/41278

### CODE REVIEW

File: xml/System.IO/Path.xml [hunk 1]
```diff
@@ -1459,7 +1459,7 @@ Unlike the string overload, this method doesn't normalize directory separators.
       <Docs>
         <param name="path">The path from which to obtain root directory information.</param>
         <summary>Gets the root directory information of the specified path.</summary>
-        <returns>The root directory of <paramref name="path" />, or <see langword="null" /> if <paramref name="path" /> is <see langword="null" />, or an empty string if <paramref name="path" /> does not contain root directory information.</returns>
+        <returns>The root directory of <paramref name="path" />, or <see langword="null" /> if <paramref name="path" /> is <see langword="null" /> or only contains spaces, or an empty string if <paramref name="path" /> does not contain root directory information.</returns>
```
Comment (stephentoub) [inline xml/System.IO/Path.xml]: The "or only contains spaces" is specific to Windows, no?

Comment (JeremyKuhne) [reply xml/System.IO/Path.xml]: Yes, that is correct. Only Windows.

Comment (carlossanlop) [reply xml/System.IO/Path.xml]: I will make sure to clarify that as well.

**Comment by carlossanlop** [comment]
@JeremyKuhne @stephentoub based on my findings when testing the GetPathRoot methods (both the string and ROS overloads), I modified their documentation and removed the exception mention. Can you please take a look again when you get a chance?

File: xml/System.IO/Path.xml [hunk 2]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
         <Parameter Name="path" Type="System.ReadOnlySpan&lt;System.Char&gt;" Index="0" FrameworkAlternate="netcore-2.1;netcore-2.2;netcore-3.0;netstandard-2.1" />
       </Parameters>
       <Docs>
-        <param name="path">The path from which to obtain root directory information.</param>
+        <param name="path">A read-only span of characters containing the path from which to obtain root directory information.</param>
         <summary>Gets the root directory information from the path contained in the specified character span.</summary>
-        <returns>A character span containing the root directory of `path`.</returns>
+        <returns>A read-only span of characters containing the root directory of `path`.</returns>
```
Comment (carlossanlop) [inline xml/System.IO/Path.xml]:
```suggestion
        <returns>A read-only span of characters containing the root directory of <paramref name="path" />.</returns>
```

File: xml/System.IO/Path.xml [hunk 3]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+
+- "X:\" (Windows: `path` specified an absolute path on a given drive).
+
+- "\\\ComputerName\SharedFolder" (Windows: a UNC path).
+
+- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions and in .NET Framework 4.6.2 and later versions)
+ 
+For more information on file paths on Windows, see [File path formats on Windows systems](~/docs/standard/io/file-path-formats.md). For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common-i-o-tasks.md).
+
+## Examples
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]: The examples section should be deleted, since it illustrates the wrong overload.

File: xml/System.IO/Path.xml [hunk 4]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+        <param name="path">A read-only span of characters containing the path from which to obtain root directory information.</param>
         <summary>Gets the root directory information from the path contained in the specified character span.</summary>
-        <returns>A character span containing the root directory of `path`.</returns>
+        <returns>A read-only span of characters containing the root directory of <paramref name="path" />.</returns>
         <remarks>
           <format type="text/markdown"><![CDATA[
 
 ## Remarks
 
+This method does not verify that the path or file name exists.
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
This method does not verify that the path or file exists.
```

File: xml/System.IO/Path.xml [hunk 5]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+
 Unlike the string overload, this method doesn't normalize directory separators.
 
+A `ReadOnlySpan<System.Char>` is "effectively empty" if:
+
+- In Windows, calling `IsEmpty` on this span of characters returns `true`, or all its characters are spaces (' ').
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
- In Windows, calling <xref:System.ReadOnlySpan%7BSystem.Char%7D.IsEmpty?displayProperty=nameWithType> on this span of characters returns `true`, or all its characters are spaces (' ').
```
I think the xref will work.

File: xml/System.IO/Path.xml [hunk 6]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+- In Unix, calling `IsEmpty` on this span of characters returns `true`.
+
+Possible patterns for the string returned by this method are as follows:
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
Possible patterns for the read-only character span returned by this method are as follows:
```

File: xml/System.IO/Path.xml [hunk 7]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+
+- `null` (`path` was null or an empty string).
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
- <xref:System.ReadOnlySpan%7BSystem.Char%7D.Empty?displayProperty=nameWithType> (`path` was <xref:System.ReadOnlySpan%7BSystem.Char%7D.Empty?displayProperty=nameWithType>.
```

File: xml/System.IO/Path.xml [hunk 8]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
 ## Remarks
 
+This method does not verify that the path or file name exists.
+
 Unlike the string overload, this method doesn't normalize directory separators.
 
+A `ReadOnlySpan<System.Char>` is "effectively empty" if:
+
+- In Windows, calling `IsEmpty` on this span of characters returns `true`, or all its characters are spaces (' ').
+- In Unix, calling `IsEmpty` on this span of characters returns `true`.
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
- In Unix, calling <xref:System.ReadOnlySpan%7BSystem.Char%7D.IsEmpty?displayProperty=nameWithType> on this span of characters returns `true`.
```

File: xml/System.IO/Path.xml [hunk 9]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+
+Possible patterns for the string returned by this method are as follows:
+
+- `null` (`path` was null or an empty string).
+
+- An empty string (`path` specified a relative path on the current drive or volume).
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
- <xref:System.ReadOnlySpan%7BSystem.Char%7D.Empty?displayProperty=nameWithType> (`path` specified a relative path on the current drive or volume).
```

File: xml/System.IO/Path.xml [hunk 10]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+
+- "\" (Unix: `path` specified an absolute path on the current drive).
+
+- "X:" (Windows: `path` specified a relative path on a drive, where X represents a drive or volume letter).
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
- "X:" (Windows: `path` specified a relative path on a drive, where *X* represents a drive or volume letter).
```

File: xml/System.IO/Path.xml [hunk 11]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+
+- "X:\" (Windows: `path` specified an absolute path on a given drive).
+
+- "\\\ComputerName\SharedFolder" (Windows: a UNC path).
+
+- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions and in .NET Framework 4.6.2 and later versions)
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions, and in .NET Framework 4.6.2 and later versions).
```

File: xml/System.IO/Path.xml [hunk 12]
```diff
@@ -1457,46 +1489,58 @@ Unlike the string overload, this method doesn't normalize directory separators.
...
- [!code-cpp[System.IO.Path Members#8](~/samples/snippets/cpp/VS_Snippets_CLR_System/system.IO.Path Members/CPP/pathmembers.cpp#8)]
- [!code-csharp[System.IO.Path Members#8](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.IO.Path Members/CS/pathmembers.cs#8)]
- [!code-vb[System.IO.Path Members#8](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.IO.Path Members/VB/pathmembers.vb#8)]  
-  
- ]]></format>
+          <format type="text/markdown"><![CDATA[
+
+## Remarks
+
+This method does not verify that the path or file name exists.
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
This method does not verify that the path or file exists.
```

File: xml/System.IO/Path.xml [hunk 13]
```diff
@@ -1457,46 +1489,58 @@ Unlike the string overload, this method doesn't normalize directory separators.
...
+
+Possible patterns for the string returned by this method are as follows:
+
+- `null` (`path` was null or an empty string).
+
+- An empty string (`path` specified a relative path on the current drive or volume).
+
+- "\" (Unix: `path` specified an absolute path on the current drive).
+
+- "X:" (Windows: `path` specified a relative path on a drive, where X represents a drive or volume letter).
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
- "X:" (Windows: `path` specified a relative path on a drive, where *X* represents a drive or volume letter).
```

File: xml/System.IO/Path.xml [hunk 14]
```diff
@@ -1457,46 +1489,58 @@ Unlike the string overload, this method doesn't normalize directory separators.
...
+
+- "X:\" (Windows: `path` specified an absolute path on a given drive).
+
+- "\\\ComputerName\SharedFolder" (Windows: a UNC path).
+
+- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions and in .NET Framework 4.6.2 and later versions)
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]:
```suggestion
- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions, and in .NET Framework 4.6.2 and later versions)
```

File: xml/System.IO/Path.xml [hunk 15]
```diff
@@ -1457,46 +1489,58 @@ Unlike the string overload, this method doesn't normalize directory separators.
...
+The following example demonstrates a use of the `GetPathRoot` method.
+
+[!code-cpp[System.IO.Path Members#8](~/samples/snippets/cpp/VS_Snippets_CLR_System/system.IO.Path Members/CPP/pathmembers.cpp#8)]
+[!code-csharp[System.IO.Path Members#8](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.IO.Path Members/CS/pathmembers.cs#8)]
+[!code-vb[System.IO.Path Members#8](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.IO.Path Members/VB/pathmembers.vb#8)]
+
+          ]]></format>
         </remarks>
-        <exception cref="T:System.ArgumentException">
-          <paramref name="path" /> contains one or more of the invalid characters defined in <see cref="M:System.IO.Path.GetInvalidPathChars" />.
```
Comment (rpetrusha) [inline xml/System.IO/Path.xml]: Does the method not throw if `path` contains invalid characters?

File: xml/System.IO/Path.xml (hunk 5 above)
Comment (carlossanlop) [reply xml/System.IO/Path.xml]: Cross fingers

File: xml/System.IO/Path.xml [hunk 16]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions and in .NET Framework 4.6.2 and later versions)
+ 
+For more information on file paths on Windows, see [File path formats on Windows systems](~/docs/standard/io/file-path-formats.md). For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common-i-o-tasks.md).
+
+## Examples
+The following example demonstrates a use of the `GetPathRoot` method.
+
+[!code-cpp[System.IO.Path Members#8](~/samples/snippets/cpp/VS_Snippets_CLR_System/system.IO.Path Members/CPP/pathmembers.cpp#8)]
+[!code-csharp[System.IO.Path Members#8](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.IO.Path Members/CS/pathmembers.cs#8)]
+[!code-vb[System.IO.Path Members#8](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.IO.Path Members/VB/pathmembers.vb#8)]
```
Comment (carlossanlop) [inline xml/System.IO/Path.xml]:
```suggestion
```

File: xml/System.IO/Path.xml [hunk 17]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+
+- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions and in .NET Framework 4.6.2 and later versions)
+ 
+For more information on file paths on Windows, see [File path formats on Windows systems](~/docs/standard/io/file-path-formats.md). For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common-i-o-tasks.md).
+
+## Examples
+The following example demonstrates a use of the `GetPathRoot` method.
+
+[!code-cpp[System.IO.Path Members#8](~/samples/snippets/cpp/VS_Snippets_CLR_System/system.IO.Path Members/CPP/pathmembers.cpp#8)]
+[!code-csharp[System.IO.Path Members#8](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.IO.Path Members/CS/pathmembers.cs#8)]
```
Comment (carlossanlop) [inline xml/System.IO/Path.xml]:
```suggestion
```

File: xml/System.IO/Path.xml [hunk 18]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+- "\\\ComputerName\SharedFolder" (Windows: a UNC path).
+
+- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions and in .NET Framework 4.6.2 and later versions)
+ 
+For more information on file paths on Windows, see [File path formats on Windows systems](~/docs/standard/io/file-path-formats.md). For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common-i-o-tasks.md).
+
+## Examples
+The following example demonstrates a use of the `GetPathRoot` method.
+
+[!code-cpp[System.IO.Path Members#8](~/samples/snippets/cpp/VS_Snippets_CLR_System/system.IO.Path Members/CPP/pathmembers.cpp#8)]
```
Comment (carlossanlop) [inline xml/System.IO/Path.xml]:
```suggestion
```

File: xml/System.IO/Path.xml [hunk 19]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+
+- "\\\ComputerName\SharedFolder" (Windows: a UNC path).
+
+- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions and in .NET Framework 4.6.2 and later versions)
+ 
+For more information on file paths on Windows, see [File path formats on Windows systems](~/docs/standard/io/file-path-formats.md). For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common-i-o-tasks.md).
+
+## Examples
+The following example demonstrates a use of the `GetPathRoot` method.
+
```
Comment (carlossanlop) [inline xml/System.IO/Path.xml]:
```suggestion
```

File: xml/System.IO/Path.xml [hunk 20]
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
...
+- "X:\" (Windows: `path` specified an absolute path on a given drive).
+
+- "\\\ComputerName\SharedFolder" (Windows: a UNC path).
+
+- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions and in .NET Framework 4.6.2 and later versions)
+ 
+For more information on file paths on Windows, see [File path formats on Windows systems](~/docs/standard/io/file-path-formats.md). For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common-i-o-tasks.md).
+
+## Examples
+The following example demonstrates a use of the `GetPathRoot` method.
```
Comment (carlossanlop) [inline xml/System.IO/Path.xml]:
```suggestion
```

File: xml/System.IO/Path.xml (hunk 3 above)
Comment (carlossanlop) [reply xml/System.IO/Path.xml]:
```suggestion
```

File: xml/System.IO/Path.xml (hunk 15 above)
Comment (carlossanlop) [reply xml/System.IO/Path.xml]: That's what I discovered on .NET Core 3.0. Let me test this again but on Framework, I suspect this comment applies only for that version.

Comment (carlossanlop) [reply xml/System.IO/Path.xml]: @rpetrusha confirmed:
- In 3.0, we don't throw when we pass "`\0`", `null` or `string.Empty`.
    - If we pass `null` or `string.Empty`, `GetPathRoot` returns `null`.
    - If we pass "`\0`", `GetPathRoot` returns `string.Empty`.
- In Framework, we throw if we pass "`\0`" or `string.Empty`. We don't throw if we pass `null`.

@JeremyKuhne is this a breaking change? Should I open an issue to bring back the exception in .NET Core for those cases?

Comment (carlossanlop) [reply xml/System.IO/Path.xml]: @rpetrusha I brought back the exception, but made sure to mention it is only thrown in .NET Framework (for both root causes).

Comment (JeremyKuhne) [reply xml/System.IO/Path.xml]: It was a deliberate breaking change in 2.1. Partially to align with the Span overloads (there are no null Spans), partially to align better with the other APIs and user expectations.

File: xml/System.IO/Path.xml [hunk 21]
```diff
@@ -1457,47 +1483,64 @@ Unlike the string overload, this method doesn't normalize directory separators.
...
+          ]]></format>
         </remarks>
-        <exception cref="T:System.ArgumentException">
-          <paramref name="path" /> contains one or more of the invalid characters defined in <see cref="M:System.IO.Path.GetInvalidPathChars" />.  
-  
- -or-  
-  
- <see cref="F:System.String.Empty" /> was passed to <paramref name="path" />.</exception>
-        <related type="article" href="~/docs/standard/io/file-path-formats.md">File path formats on Windows systems</related>
+        <exception cref="T:System.ArgumentException"><paramref name="path" />.NET Framework only: contains one or more of the invalid characters defined in <see cref="M:System.IO.Path.GetInvalidPathChars" />.
```
Comment (mairaw) [inline xml/System.IO/Path.xml]: should this be like this?

```suggestion
        <exception cref="T:System.ArgumentException">.NET Framework only: <paramref name="path" /> contains one or more of the invalid characters defined in <see cref="M:System.IO.Path.GetInvalidPathChars" />.
```

**Comment by carlossanlop** [comment]
@mairaw I think this one's fully cooked unless you say otherwise.

**Comment by carlossanlop** [comment]
@mairaw I addressed the comments, and Ron approved it. Can you please merge this whenever you get a chance?

**Pull request merged**